src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapperUtil.java

```diff
@@ -237,37 +223,184 @@ public static Object deserializeStoredVector(BytesRef storedVector, VectorDataTy
         }
         return VectorDataType.BINARY == knnVectorFieldType.getVectorDataType() ? expectedDimensions / 8 : expectedDimensions;
     }
 
-    private static boolean isModelBasedIndex(int expectedDimensions) {
-        return expectedDimensions == -1;
-    }
-
     /**
      * Returns the model metadata for a specified knn vector field
      *
      * @param knnVectorField knn vector field
      * @return the model metadata from knnVectorField
      */
     private static ModelMetadata getModelMetadataForField(final KNNVectorFieldType knnVectorField) {
-        String modelId = knnVectorField.getModelId();
+        ModelMetadata[] modelMetadata = new ModelMetadata[1];
+        knnVectorField.getAnnConfig().getModelId().ifPresentOrElse(modelId -> {
+            modelMetadata[0] = modelDao.getMetadata(modelId);
+            if (!ModelUtil.isModelCreated(modelMetadata[0])) {
+                throw new IllegalArgumentException(String.format("Model ID '%s' is not created.", modelId));
+            }
+        }, () -> { throw new IllegalArgumentException(String.format("Field '%s' does not have model.", knnVectorField.name())); });
+        return modelMetadata[0];
+    }
 
-        if (modelId == null) {
-            throw new IllegalArgumentException(
-                String.format("Field '%s' does not have model.", knnVectorField.getKnnMethodContext().getMethodComponentContext().getName())
+    /**
+     * Validate if the circuit breaker is triggered
+     */
+    public static void validateIfCircuitBreakerIsNotTriggered() {
+        if (KNNSettings.isCircuitBreakerTriggered()) {
+            throw new KnnCircuitBreakerException(
+                "Parsing the created knn vector fields prior to indexing has failed as the circuit breaker triggered.  This indicates that the cluster is low on memory resources and cannot index more documents at the moment. Check _plugins/_knn/stats for the circuit breaker status."
+            );
+        }
+    }
+
+    /**
+     * Validate if plugin is enabled
+     */
+    public static void validateIfKNNPluginEnabled() {
+        if (!KNNSettings.isKNNPluginEnabled()) {
+            throw new IllegalStateException("KNN plugin is disabled. To enable update knn.plugin.enabled setting to true");
+        }
+    }
+
+    /**
+     * Get space type or default based on settings and data type
+     *
+     * @param indexSettings Settings
+     * @param vectorDataType VectorDataType
+     * @return SpaceType as string
+     */
+    public static SpaceType getSpaceType(final Settings indexSettings, final VectorDataType vectorDataType) {
+        String spaceType = indexSettings.get(KNNSettings.INDEX_KNN_SPACE_TYPE.getKey());
+        if (spaceType == null) {
+            spaceType = VectorDataType.BINARY == vectorDataType
+                ? KNNSettings.INDEX_KNN_DEFAULT_SPACE_TYPE_FOR_BINARY
+                : KNNSettings.INDEX_KNN_DEFAULT_SPACE_TYPE;
+            log.info(
```

Comment:
You don't need String.format here.
```suggestion
            log.info("[KNN] The setting \"{}\" was not set for the index. Likely caused by recent version upgrade. Setting the setting to the default value={}",
                    METHOD_PARAMETER_SPACE_TYPE,
                    spaceType);
```

Reply:
For copy and pasted changes, I would prefer to not alter in this PR just to limit scope. It can be followed up in another PR.